### PR TITLE
Initial conditions

### PR DESCRIPTION
**Pull request by vkrajnak**
Functions that accommodate calculation of initial conditions on various sections and in various systems.

### CODE REVIEW

**Comment by vkrajnak** [comment]
I should add that there is a discrepancy in the definitions of our systems. While vector fields accept  (t,u) as input, potentials cannot currently be time-dependent. This can be addressed in a separate pull request.

File: pylds/hamiltonians.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+    Returns
+    -------
+    Hamiltonian : function
+        Sum of 'potential' and 'kinetic_squares'.
+
+    """
+    def Hamiltonian(t,u):
+        return potential(u) + kinetic_squares(t,u)
+    return Hamiltonian
```
Comment (broncio123) [inline pylds/hamiltonians.py:255]: WRONG INPUT VARIABLE for `potential`.
Here, `u` is taken as an array of phase space points. The function `kinetic_squares` correctly takes points corresponding to conjugate momenta axes (`momenta_points`) from `u`, but all potential energy functions only accept `positions`, leading to a `ValueError` when running `NFSaddle` example.

**Review by broncio123 — CHANGES_REQUESTED** [review]
Hi @vkrajnak 

I just finished reviewing your PR.

Overall, I am happy with the new changes to the package, including
* redefinition of the function `generate_points`, and the
* addition of `hamiltonians.py` module

But,
* docstrings and comments for some methods are missing or are required to clarify the logic of some code blocks.

All the examples in the Examples NB run as they should, except for the Index-1 Normal Form saddle.

You would like to check my comment on the function `Hamiltonian_from_potential`, to correct the input variables for `potential` which must be `positions`. 

After the above points are addressed you can go ahead with merging.

mad love,
b

File: pylds/base.py [hunk 2]
```diff
@@ -136,7 +163,7 @@ def pass_axis_parameters(parameters): return np.linspace(*parameters)
             points_y = np.linspace(y_min, y_max, Ny)
             X, Y = np.meshgrid(points_x, points_y)  # Grid in phase space
             # 2D grid + a zero column for LDs
-            mesh = np.transpose([X.flatten(), Y.flatten(), np.zeros(Nx*Ny)])
+            mesh = np.transpose([X.ravel(), Y.ravel(), np.zeros(Nx*Ny)])
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/base.py:166]: Flatten returns a copy of the array, ravel returns the same pointer with rearranged elements. Should save a tiny bit of memory and time

File: pylds/base.py [hunk 3]
```diff
@@ -37,91 +49,106 @@ def generate_points(grid_parameters):
...
+                phase_space_axes, H0, Hamiltonian, momentum_sign)
+        except:
+            remaining_coordinate_bounds = np.array(grid_parameters['remaining_coordinate_bounds'])
+            def f_remaining_coordinate_value(u):
+                return remaining_coordinate_value(u, ind_remaining, remaining_coordinate_bounds, H0, Hamiltonian)
+            phase_space_axes[:,ind_remaining] = np.array(list(    \
+                            map(f_remaining_coordinate_value, phase_space_axes)))
+
+        mask = np.isnan(phase_space_axes[:,ind_remaining]) # Mask grid points
+        phase_space_axes[:,ind_remaining] = np.nan_to_num(phase_space_axes[:,ind_remaining])
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/base.py]: You got it, I'll comment on it

File: pylds/hamiltonians.py (hunk 1 above)
Comment (vkrajnak) [reply pylds/hamiltonians.py:255]: Thanks for finding this issue!

File: pylds/hamiltonians.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+    Returns
+    -------
+    H : array_like, shape(n,)
+        Hamiltonian at points u, in phase space at time t
+    """
+    x, y = u.T
+    # Hamiltonian Model Parameter
+    omega, = PARAMETERS
+    return 0.5*omega*(y*y + x*x)
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/hamiltonians.py:35]: Are you suggesting to change the notation in all 1DoF systems, maps and vector fields? If so, I'd suggest a separate issue/pull request, to keep things organised

File: pylds/hamiltonians.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    u : array_like, shape(n,)
+        Points in phase space.
+
+    Returns
+    -------
+    T : array_like, shape(n,)
+        Kinetic energy at points u, in phase space at time t
+
+    """
+    N_dim = np.floor(u.shape[1]/2).astype('int')
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/hamiltonians.py]: Done

File: pylds/base.py [hunk 6]
```diff
@@ -10,20 +10,32 @@
...
-    points_dims_remaining = momentum_sign * \
-        np.sqrt(2*(H0 - V) - (phase_space_momenta**2).sum(axis=1))
+def remaining_coordinate_quadratic(phase_space_axes, H0, Hamiltonian, momentum_sign):
+    # phase_space_axes[:, ind_remaining] initialised at minimum: 0
+
+    energy = H0-Hamiltonian(0, phase_space_axes)
+    masked_energy = np.copy(energy)
+    masked_energy[np.argwhere(energy<0)]=np.nan
+    points_dims_remaining = momentum_sign * np.sqrt(2*masked_energy)
     return points_dims_remaining
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/base.py]: Done

File: pylds/base.py [hunk 7]
```diff
@@ -10,20 +10,32 @@
...
+
+    def remaining_energy(guess):
+        u[ind_remaining] = guess
+        return H0 - Hamiltonian(0, u)
+    try:
+        points_dims_remaining = brentq(remaining_energy, remaining_coordinate_bounds[0], remaining_coordinate_bounds[1])
+        return points_dims_remaining
+    except:
+        return np.nan
+
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/base.py]: Done

File: pylds/base.py [hunk 8]
```diff
@@ -37,91 +49,106 @@ def generate_points(grid_parameters):
...
+        dims_fixed_values = np.array(grid_parameters['dims_fixed_values'])
         H0 = grid_parameters['energy_level']
 
+        try:
+            Hamiltonian = grid_parameters['Hamiltonian']
+        except:
+            potential_energy = grid_parameters['potential_energy']
+            Hamiltonian = Hamiltonian_from_potential(potential_energy)
+
+
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/base.py:132]: Added comment

File: pylds/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/base.py]: Added comment

File: pylds/hamiltonians.py (hunk 1 above)
Comment (vkrajnak) [reply pylds/hamiltonians.py]: Fixed

**Comment by vkrajnak** [comment]
Fixed, added, updated

**Comment by vkrajnak** [comment]
Let me know if you spot anything else

File: pylds/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (broncio123) [reply pylds/base.py:219]: Excellent!

**Review by broncio123 — APPROVED** [review]
I reviewed all changes and comments.
I also executed Examples NB and now everything works fine.
PR should go ahead.

**Comment by broncio123** [comment]
I'll take care of the merging conflicts myself.

**Pull request merged**